### ISSUE TITLE
High memory usage when consistently using new HTTP client instances for each request

### ISSUE THREAD

**Issue by nizacegodk** [guzzle/guzzle#3162]
**Guzzle version(s) affected**: 7.7.0
**PHP version**: 8.2.7
**cURL version**: 7.88.1

**Description**  

I have a long running PHP script deployed within a docker service, which over its lifetime sends many HTTP requests. 
I have noticed that the memory usage reported by `docker stats` keep increasing and settles at around 2GB of used memory - If I remove the line sending the HTTP requests, then the memory stays low and stable.

Inspection with `memory_get_usage()` in the PHP runtime, does report a significantly lower memory usage than `docker stats` does.
I can reproduce a similar memory gap between the two tools, by simply creating multiple curl handles and never closing them - Leaving me to the thinking that the high memory usage is caused by hanging curl handles?

If I create a single guzzle client instance and reuse that client for all requests, then I do not see any large memory usage. 

**How to reproduce**  

Deploy at docker service with the following script
```
while (true) {
    $guzzle = new Client();

    try {
        $guzzle->get('https://some_url_you_do_not_mind_spamming');
    } catch (Throwable) {
        // Do nothing
    }
}
```

And inspect the memory usage with `docker stats`

The host system is Ubuntu 22.

**Possible Solution**  

It might be related to some curl handles not being closed?

**Additional context**  

I have attempted calling `gc_collect_cycles()` within my script, which does nothing for the high memory.
The high memory usage can also be measured on the host system using tools such as `btop` to measure the difference between "Used memory", "Available memory", "Cached memory" and "Free memory". Killing the docker service after it reaches the 2 GB peak, I can see a similar drop in "Used memory". No affect is present on "Cached memory".

**Comment by nizacegodk** [comment]
I did some additional investigation.

If I modify the code of the `CurlFactory::release()` method to always close the curl handle after each request, and thereby not use the `maxHandles` logic, then the memory settles way lower - At around 150MB.

If I add a destruct method to `CurlFactory::class` with the following implementation, closing all handles on object destruction, then the memory seams to settle at around 250 MB.

I currently cannot explain the 100MB difference between closing the curl handles immediately and on object destruction, but the issue seams to be a missing class destructor automatically closing curl handles once the client is no longer in use. 

```php
public function __destruct()
{
    foreach ($this->handles as $id => $handle) {
        curl_close($handle);
        unset($this->handles[$id]);
    }
}
```

**Comment by nizacegodk** [comment]
I have submitted a PR for the suggested solution at https://github.com/guzzle/guzzle/pull/3165

**Comment by GrahamCampbell** [comment]
Fixed in v7.8.0.
